Code/Framework/AzCore/AzCore/IO/Path/PathParser.inl

```diff
@@ -0,0 +1,706 @@
+                return AZStd::next(entryBeginIter, 2);
+            }
+
+            if (!Internal::IsSeparator(path[0]))
+            {
+                // At this point all other root names start with a path separator
+                return entryBeginIter;
+            }
+
+            // Check if the path has the form of "\\?\, "\??\" or "\\.\"
+            const bool pathInUncForm = path.size() >= 4 && Internal::IsSeparator(path[3])
+                && (path.size() == 4 || !Internal::IsSeparator(path[4]));
+            if (pathInUncForm)
+            {
+                // \\?\<0 or more> or \\.\$<zero or more>
+                const bool slashQuestionMark = Internal::IsSeparator(path[1]) && (path[2] == '?' || path[2] == '.');
+                // \??\<0 or more>
+                const bool questionMarkTwice = path[1] == '?' && path[2] == '?';
+                if (slashQuestionMark || questionMarkTwice)
+                {
+                    // Return the root value root slash - i.e "\\?"
+                    return AZStd::next(entryBeginIter, 3);
+                }
+            }
+
+            if (path.size() >= 3 && Internal::IsSeparator(path[1]) && !Internal::IsSeparator(path[2]))
+            {
+                // Find the next path separator for network paths that have the form of \\server\share
+                constexpr AZStd::string_view PathSeparators = { "/\\" };
+                size_t nextPathSeparatorOffset = path.find_first_of(PathSeparators, 3);
+                return AZStd::next(entryBeginIter, nextPathSeparatorOffset != AZStd::string_view::npos ? nextPathSeparatorOffset : path.size());
+            }
+
+            return entryBeginIter;
+        }
+    }
+
+    //! Returns an iterator past the end of the consumed path separator(s)
+    template <typename InputIt>
+    constexpr InputIt ConsumeSeparator(InputIt entryBeginIter, InputIt entryEndIter) noexcept
+    {
+        return AZStd::find_if_not(entryBeginIter, entryEndIter, [](const char elem) { return Internal::IsSeparator(elem); });
+    }
+
+    //! Returns an iterator past the end of the consumed filename
+    template <typename InputIt>
+    constexpr InputIt ConsumeName(InputIt entryBeginIter, InputIt entryEndIter) noexcept
+    {
+        return AZStd::find_if(entryBeginIter, entryEndIter, [](const char elem) { return Internal::IsSeparator(elem); });
+    }
+
+    //! Check if a path is absolute on a OS basis
+    //! If the preferred separator is '/' just checks if the path starts with a '/
+    //! Otherwise a check for a Windows absolute path occurs
+    //! Windows absolute paths can include a RootName
+    template <typename InputIt, typename EndIt, typename = AZStd::enable_if_t<AZStd::Internal::is_input_iterator_v<InputIt>>>
+    static constexpr bool IsAbsolute(InputIt first, EndIt last, const char preferredSeparator)
+    {
+        size_t pathSize = AZStd::distance(first, last);
```

Comment:
When AZ_TRAIT_CUSTOM_PATH_ROOT_SEPARATOR is defined, this results in an unused local variable warning (that is treated as an error). In https://github.com/o3de/o3de/pull/3678 I just moved it to be calculated right before it's used in the two places below.

Reply:
It is not an unused variable, because the else block always uses it.

Reply:
Ah true, I hit that when I originally had the #if check at the function scope but the pathSize outside of it. Never mind me...